In [13]:
import pandas as pd
import numpy as np

import time, warnings
import datetime as dt
warnings.filterwarnings("ignore")

In [14]:
#loading dataset
retail_df = pd.read_excel("E:\\PGP-BABI\\Capstone Project\\Dataset\\Online Retail II.xlsx")
retail_df.head(10)

Invoice StockCode                          Description  Quantity  \
0  536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1  536365     71053                  WHITE METAL LANTERN         6   
2  536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3  536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4  536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
5  536365     22752         SET 7 BABUSHKA NESTING BOXES         2   
6  536365     21730    GLASS STAR FROSTED T-LIGHT HOLDER         6   
7  536366     22633               HAND WARMER UNION JACK         6   
8  536366     22632            HAND WARMER RED POLKA DOT         6   
9  536368     22960             JAM MAKING SET WITH JARS         6   

          InvoiceDate  Price  Customer ID         Country  
0 2010-12-01 08:26:00   2.55      17850.0  United Kingdom  
1 2010-12-01 08:26:00   3.39      17850.0  United Kingdom  
2 2010-12-01 08:26:00   2.75      17850.0  United Kingdom  
3 2010-12-01 08:26:00   3.39      17850.0  United Kingdom  
4 2010-12-01 08:26:00   3.39      17850.0  United Kingdom  
5 2010-12-01 08:26:00   7.65      17850.0  United Kingdom  
6 2010-12-01 08:26:00   4.25      17850.0  United Kingdom  
7 2010-12-01 08:28:00   1.85      17850.0  United Kingdom  
8 2010-12-01 08:28:00   1.85      17850.0  United Kingdom  
9 2010-12-01 08:34:00   4.25      13047.0  United Kingdom

In [15]:
retail_df.tail(10)

Invoice StockCode                      Description  Quantity  \
541900  581587     22730       ALARM CLOCK BAKELIKE IVORY         4   
541901  581587     22367  CHILDRENS APRON SPACEBOY DESIGN         8   
541902  581587     22629              SPACEBOY LUNCH BOX         12   
541903  581587     23256      CHILDRENS CUTLERY SPACEBOY          4   
541904  581587     22613      PACK OF 20 SPACEBOY NAPKINS        12   
541905  581587     22899     CHILDREN'S APRON DOLLY GIRL          6   
541906  581587     23254    CHILDRENS CUTLERY DOLLY GIRL          4   
541907  581587     23255  CHILDRENS CUTLERY CIRCUS PARADE         4   
541908  581587     22138    BAKING SET 9 PIECE RETROSPOT          3   
541909  581587      POST                          POSTAGE         1   

               InvoiceDate  Price  Customer ID Country  
541900 2011-12-09 12:50:00   3.75      12680.0  France  
541901 2011-12-09 12:50:00   1.95      12680.0  France  
541902 2011-12-09 12:50:00   1.95      12680.0  France  
541903 2011-12-09 12:50:00   4.15      12680.0  France  
541904 2011-12-09 12:50:00   0.85      12680.0  France  
541905 2011-12-09 12:50:00   2.10      12680.0  France  
541906 2011-12-09 12:50:00   4.15      12680.0  France  
541907 2011-12-09 12:50:00   4.15      12680.0  France  
541908 2011-12-09 12:50:00   4.95      12680.0  France  
541909 2011-12-09 12:50:00  18.00      12680.0  France

In [16]:
# Data preparation
retail_uk = retail_df[retail_df['Country']=='United Kingdom']
# checking the shape
retail_uk.shape

(495478, 8)

In [17]:
# Removing cancelled orders
retail_uk = retail_uk[retail_uk['Quantity']>0]
retail_uk.shape

(486286, 8)

In [18]:
# Removing rows where customerID is NA
retail_uk.dropna(subset=['Customer ID'],how='all',inplace=True)
retail_uk.shape

(354345, 8)

In [19]:
# Restrict the data to one full year because it's better to use a metric per Months or Years in RFM
retail_uk = retail_uk[retail_uk['InvoiceDate']>= "2010-12-09"]
retail_uk.shape

(342478, 8)

In [20]:
print("Summary:")
#exploring the unique values of each attribute
print("Number of transactions: ", retail_uk['Invoice'].nunique())
print("Number of products bought: ",retail_uk['StockCode'].nunique())
print("Number of customers:", retail_uk['Customer ID'].nunique() )
print("Percentage of customers NA: ", round(retail_uk['Customer ID'].isnull().sum() * 100 / len(retail_df),2),"%" )

Summary:
Number of transactions:  16017
Number of products bought:  3611
Number of customers: 3863
Percentage of customers NA:  0.0 %


In [21]:
# RFM Analysis
# Recency
# last date available in our dataset
retail_uk['InvoiceDate'].max()

Timestamp('2011-12-09 12:49:00')

In [34]:
retail_uk['date'] = retail_uk['InvoiceDate'].dt.date

In [22]:
# since, it's 2011 dataset we have maintained the same period for analysis
now = dt.date(2011,12,9)
print(now)

2011-12-09


In [35]:
# Group by customers and check last date of purshace
recency_df = retail_uk.groupby(by='Customer ID', as_index=False)['date'].max()
recency_df.columns = ['Customer ID','LastPurshaceDate']
recency_df.head()

Customer ID LastPurshaceDate
0      12346.0       2011-01-18
1      12747.0       2011-12-07
2      12748.0       2011-12-09
3      12749.0       2011-12-06
4      12820.0       2011-12-06

In [36]:
# Calculate recency
recency_df['Recency'] = recency_df['LastPurshaceDate'].apply(lambda x: (now - x).days)

In [37]:
recency_df.head()

Customer ID LastPurshaceDate  Recency
0      12346.0       2011-01-18      325
1      12747.0       2011-12-07        2
2      12748.0       2011-12-09        0
3      12749.0       2011-12-06        3
4      12820.0       2011-12-06        3

In [38]:
recency_df.drop('LastPurshaceDate',axis=1,inplace=True)

# Frequency

In [40]:
# drop duplicates
retail_uk_copy = retail_uk
retail_uk_copy.drop_duplicates(subset=['Invoice', 'Customer ID'], keep="first", inplace=True)
#calculate frequency of purchases
frequency_df = retail_uk_copy.groupby(by=['Customer ID'], as_index=False)['Invoice'].count()
frequency_df.columns = ['Customer ID','Frequency']
frequency_df.head()

Customer ID  Frequency
0      12346.0          1
1      12747.0         10
2      12748.0        196
3      12749.0          5
4      12820.0          4

# Monetary

In [43]:
#create column total cost
retail_uk['TotalCost'] = retail_uk['Quantity'] * retail_uk['Price']

In [44]:
monetary_df = retail_uk.groupby(by='Customer ID',as_index=False).agg({'TotalCost': 'sum'})
monetary_df.columns = ['Customer ID','Monetary']
monetary_df.head()

Customer ID  Monetary
0      12346.0  77183.60
1      12747.0    658.89
2      12748.0   3739.23
3      12749.0     98.35
4      12820.0     58.20

# Create RFM Table

In [45]:
#merge recency dataframe with frequency dataframe
temp_df = recency_df.merge(frequency_df,on='Customer ID')
temp_df.head()

Customer ID  Recency  Frequency
0      12346.0      325          1
1      12747.0        2         10
2      12748.0        0        196
3      12749.0        3          5
4      12820.0        3          4

In [47]:
#merge with monetary dataframe to get a table with the 3 columns
rfm_df = temp_df.merge(monetary_df,on='Customer ID')
#use CustomerID as index
rfm_df.set_index('Customer ID',inplace=True)
#check the head
rfm_df.head()

Recency  Frequency  Monetary
Customer ID                              
12346.0          325          1  77183.60
12747.0            2         10    658.89
12748.0            0        196   3739.23
12749.0            3          5     98.35
12820.0            3          4     58.20

Customer with ID = 12346 has recency: 325 days, frequency:1, and monetary: 77183,60 £.

# RFM Table Correctness verification

In [49]:
retail_uk[retail_uk['Customer ID']==12346.0]

Invoice StockCode                     Description  Quantity  \
61619  541431     23166  MEDIUM CERAMIC TOP STORAGE JAR     74215   

              InvoiceDate  Price  Customer ID         Country        date  \
61619 2011-01-18 10:01:00   1.04      12346.0  United Kingdom  2011-01-18   

       TotalCost  
61619    77183.6

In [50]:
(now - dt.date(2011,1,18)).days == 325

True

The customer bought only one product with a huge price